xonsh/completers/pip.py

```diff
@@ -24,33 +26,41 @@ def ALL_COMMANDS():
             subprocess.check_output(["pip", "--help"], stderr=subprocess.DEVNULL)
         )
     except FileNotFoundError:
-        return []
+        try:
+            help_text = str(
+                subprocess.check_output(["pip3", "--help"], stderr=subprocess.DEVNULL)
+            )
+        except FileNotFoundError:
+            return []
     commands = re.findall(r"  (\w+)  ", help_text)
     return [c for c in commands if c not in ["completion", "help"]]
 
 
-def complete_pip(prefix, line, begidx, endidx, ctx):
+@contextual_command_completer
+def complete_pip(context: CommandContext):
     """Completes python's package manager pip"""
-    line_len = len(line.split())
-    if (
-        (line_len > 3)
-        or (line_len > 2 and line.endswith(" "))
-        or (not PIP_RE.search(line))
-    ):
-        return
-    if PIP_LIST_RE.search(line):
+    prefix = context.prefix
+    if context.arg_index == 0 or (not PIP_RE.search(context.args[0].value)):
+        return None
+    filter_func = get_filter_function()
+
+    if context.arg_index == 2 and context.args[1].value in PIP_LIST_COMMANDS:
+        # `pip show PREFIX` - complete package names
         try:
-            items = subprocess.check_output(["pip", "list"], stderr=subprocess.DEVNULL)
+            enc_items = subprocess.check_output(
+                [context.args[0].value, "list"], stderr=subprocess.DEVNULL
+            )
         except FileNotFoundError:
-            return set()
-        items = items.decode("utf-8").splitlines()
-        return set(i.split()[0] for i in items if i.split()[0].startswith(prefix))
-
-    if (line_len > 1 and line.endswith(" ")) or line_len > 2:
-        # "pip show " -> no complete (note space)
-        return
-    if prefix not in ALL_COMMANDS:
-        suggestions = [c for c in ALL_COMMANDS if c.startswith(prefix)]
+            return None
+        packages = (
+            line.split(maxsplit=1)[0] for line in enc_items.decode("utf-8").splitlines()
+        )
+        return {package for package in packages if filter_func(package, prefix)}
```

Comment:
`pip show ` also shows me an entry for `----------` that should be caught by the `filter_func` but isn't for some reason.

Also, `pip show<tab>` isn't inserting a space, which I would expect

Reply:
These two also exist before this PR, but I've added a fix for both (and tests!)